### PR TITLE
 Implement Device as a type in the script

### PR DESCRIPTION
**Pull request by zdevito**
[ note:  stacked on expect files changes, will unstack once they land ]
This adds DeviceObjType (cannot use DeviceType it is already an enum)
to the type hierarchy and an isDevice/toDevice pair to IValue.
Previous hacks which used an int[] to represent Device are removed
and at::Device is used instead.

Note: the behavior or .to is only a subset of python, we need to
fix the aten op so that it accepts Option[Device] and Optional[ScalarType].

### CODE REVIEW

File: aten/src/ATen/core/ivalue.h [hunk 1]
```diff
@@ -367,6 +368,28 @@ struct CAFFE2_API IValue final {
...
+  }
+  bool isDevice() const { return Tag::Device == tag; }
+  c10::Device toDevice() const {
+    AT_ASSERT(isDevice());
+    return c10::Device(payload.as_device.type, payload.as_device.index);
+  }
+
+  // ScalarType
+  at::ScalarType toScalarType() const {
+    return static_cast<at::ScalarType>(toInt());
```
Comment (suo) [inline aten/src/ATen/core/ivalue.h]: assert isDevice here I think.

File: aten/src/ATen/core/ivalue.h [hunk 2]
```diff
@@ -367,6 +368,28 @@ struct CAFFE2_API IValue final {
...
+  }
+
+  // Layout
+  at::Layout toLayout() const {
+    return static_cast<at::Layout>(toInt());
```
Comment (suo) [inline aten/src/ATen/core/ivalue.h]: and here

**Review by suo — APPROVED** [review]
lgtm

File: aten/src/ATen/core/ivalue.h (hunk 1 above)
Comment (zdevito) [reply aten/src/ATen/core/ivalue.h]: `toInt` does the assert, and that is all we know (until we a add first-class ScalarType support)

**Pull request closed without merging**